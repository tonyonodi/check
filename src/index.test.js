import {
  CheckError,
  isNumber,
  isInteger,
  isString,
  isBool,
  isVal,
  isInstanceOf,
  isObjectWithShape,
  isArray,
  isArrayOf,
  maybe,
} from "./index";

describe("isNumber", () => {
  test(`passing a number returns the number`, () => {
    expect(isNumber(5)).toBe(5);
  });

  test(`passing a string throws a CheckError`, () => {
    expect(() => isNumber("s")).toThrowErrorMatchingSnapshot();
  });
});

describe("isInteger", () => {
  test(`passing an integer returns the integer`, () => {
    expect(isInteger(5)).toEqual(5);
  });

  test(`passing string throws a CheckError`, () => {
    expect(() => isInteger("f")).toThrowErrorMatchingSnapshot();
  });

  test(`passing float throws a CheckError`, () => {
    expect(() => isInteger(5.1)).toThrowErrorMatchingSnapshot();
  });
});

describe("isString", () => {
  test(`passing string returns the string`, () => {
    expect(isString("a string")).toEqual("a string");
  });

  test(`passing an object throws a CheckError`, () => {
    expect(() => isString({ a: 5 })).toThrowErrorMatchingSnapshot();
  });
});

describe("isBool", () => {
  test(`passing boolean returns the boolean`, () => {
    expect(isBool(true)).toEqual(true);
    expect(isBool(false)).toEqual(false);
  });

  test(`passing non-boolean falsey values throws a CheckError`, () => {
    let error;
    expect(() => isBool(undefined)).toThrowErrorMatchingSnapshot();
    expect(() => isBool(null)).toThrowErrorMatchingSnapshot();
    expect(() => isBool(0)).toThrowErrorMatchingSnapshot();
  });

  test(`passing non-boolean truthy values throws a CheckError`, () => {
    expect(() => isBool("a")).toThrowErrorMatchingSnapshot();
    expect(() => isBool(1)).toThrowErrorMatchingSnapshot();
    expect(() => isBool([])).toThrowErrorMatchingSnapshot();
  });
});

describe(`isArray`, () => {
  test(`passing an array to isArray returns the array`, () => {
    expect(isArray([1, 2, 3])).toEqual([1, 2, 3]);
  });

  test(`passing non-array throws a CheckError`, () => {
    expect(() => isArray(false)).toThrowErrorMatchingSnapshot();
  });
});

describe("isVal", () => {
  test(`passing number returns the number`, () => {
    expect(isVal(6)(6)).toEqual(6);
  });
  test(`passing array returns the array`, () => {
    expect(isVal([1, 2, 3, 4])([1, 2, 3, 4])).toEqual([1, 2, 3, 4]);
    expect(isVal([1, [2, 3]])([1, [2, 3]])).toEqual([1, [2, 3]]);
  });
  test(`passing object returns the object`, () => {
    expect(isVal({ a: 1, b: 2 })({ a: 1, b: 2 })).toEqual({ a: 1, b: 2 });
    expect(isVal({ a: 1, b: { c: 3 } })({ a: 1, b: { c: 3 } })).toEqual({
      a: 1,
      b: { c: 3 },
    });
  });
  test(`passing wrong number throws CheckError`, () => {
    expect(() => isVal(5)(6)).toThrowErrorMatchingSnapshot();
  });
  test(`passing wrong value when expecting array throws CheckError`, () => {
    expect(() => isVal([])(5)).toThrowErrorMatchingSnapshot();
    expect(() => isVal([1, 2, 3])([1, 2, 2])).toThrowErrorMatchingSnapshot();
  });
  test(`passing wrong value when expecting nested array throws CheckError`, () => {
    expect(() =>
      isVal([1, 2, [3]])([1, 2, [4]])
    ).toThrowErrorMatchingSnapshot();
  });
  test(`passing wrong value when expecting object throws CheckError`, () => {
    const isEmptyObject = isVal({});
    const isSimpleObject = isVal({ num: 1, str: "string" });
    const isNestedObject = isVal({ num: 1, arr: [1, "b"], obj: { a: 1 } });

    expect(() => isEmptyObject(5)).toThrowErrorMatchingSnapshot();
    expect(() => isEmptyObject({ a: 1 })).toThrowErrorMatchingSnapshot();
    expect(() => isSimpleObject({ a: 1 })).toThrowErrorMatchingSnapshot();
    expect(() =>
      isNestedObject({ num: 1, arr: [1, "b"], obj: { a: 2 } })
    ).toThrowErrorMatchingSnapshot();
  });
});

describe("isInstanceOf", () => {
  test(`passing correct value returns value`, () => {
    expect(isInstanceOf(Date)(new Date(0))).toEqual(new Date(0));
  });

  test(`passing incorrect value throws CheckError`, () => {
    expect(() => isInstanceOf(Date)(5)).toThrowErrorMatchingSnapshot();
  });
});

describe("isArrayOf", () => {
  const isArrayOfStrings = isArrayOf(isString, { name: "isArrayOfStrings" });

  test(`returns array when given correct values`, () => {
    expect(isArrayOfStrings(["a", "b", "c"])).toEqual(["a", "b", "c"]);
  });

  test(`passing non-array when expecting array throws CheckError`, () => {
    expect(() => isArrayOfStrings(5)).toThrowErrorMatchingSnapshot();
  });

  test(`passing array with element that fails test throws CheckError`, () => {
    expect(() =>
      isArrayOfStrings(["a", 2, "c"])
    ).toThrowErrorMatchingSnapshot();
  });
});

describe("maybe", () => {
  test(`passing undefined returns undefined`, () => {
    expect(maybe(isString)(undefined)).toEqual(undefined);
  });

  test(`passing value with correct type returns value`, () => {
    expect(maybe(isString)("a")).toEqual("a");
  });

  test(`passing value with incorrect type throws CheckError`, () => {
    expect(() => maybe(isString)(5)).toThrowErrorMatchingSnapshot();
  });
});

describe("isObjectWithShape", () => {
  const isPerson = isObjectWithShape(
    {
      req: {
        firstName: isString,
        surname: isString,
        dateOfBirth: isInteger,
      },
      opt: {
        maidenName: isString,
      },
    },
    { name: "isPerson" }
  );
  test(`passing object with correct shape returns object`, () => {
    expect(
      isPerson({
        firstName: "John",
        surname: "Doe",
        dateOfBirth: 1990,
      })
    ).toEqual({
      firstName: "John",
      surname: "Doe",
      dateOfBirth: 1990,
    });

    expect(
      isPerson({
        firstName: "Jane",
        surname: "Doe",
        dateOfBirth: 1990,
        maidenName: "Smith",
      })
    ).toEqual({
      firstName: "Jane",
      surname: "Doe",
      dateOfBirth: 1990,
      maidenName: "Smith",
    });
  });

  test(`passing nested object with correct shape returns nested object`, () => {
    const isCarShape = isObjectWithShape({
      req: {
        seats: isObjectWithShape({
          req: {
            number: isInteger,
            trim: isString,
          },
        }),
        doors: isInteger,
      },
      opt: {
        engine: isObjectWithShape({
          req: {
            size: isInteger,
            aspiration: isString,
            fuel: isString,
          },
        }),
      },
    });
    expect(
      isCarShape({
        seats: {
          number: 4,
          trim: "cloth",
        },
        doors: 4,
        engine: {
          size: 1900,
          aspiration: "turbocharged",
          fuel: "diesel",
        },
      })
    ).toEqual({
      seats: {
        number: 4,
        trim: "cloth",
      },
      doors: 4,
      engine: {
        size: 1900,
        aspiration: "turbocharged",
        fuel: "diesel",
      },
    });
  });

  test(`passing object missing required key throws CheckError`, () => {
    expect(() =>
      isPerson({
        firstName: "John",
        surname: "Doe",
      })
    ).toThrowErrorMatchingSnapshot();
  });

  test(`passing object with wrong property in req throws error`, () => {
    const isPerson = isObjectWithShape(
      { req: { firstName: isString } },
      { name: "isPerson" }
    );
    expect(() => isPerson({ firstName: 5 })).toThrowErrorMatchingSnapshot();
  });

  test(`passing object with wrong property in opt throws error`, () => {
    const isPerson = isObjectWithShape(
      { req: { firstName: isString }, opt: { dateOfBirth: isInteger } },
      { name: "isPerson" }
    );

    expect(() =>
      isPerson({ firstName: "Jane", dateOfBirth: 199.5 })
    ).toThrowErrorMatchingSnapshot();
  });

  test(`passing object with nested wrong property in req throws error`, () => {
    const isName = isObjectWithShape(
      {
        req: {
          firstName: isString,
          surname: isString,
        },
      },
      { name: "isName" }
    );
    const isPerson = isObjectWithShape(
      {
        req: {
          name: isName,
        },
      },
      { name: "isPerson" }
    );

    expect(() =>
      isPerson({ name: { firstName: "John", surname: 5 } })
    ).toThrowErrorMatchingSnapshot();
  });

  test(`passing object with nested missing property in req throws error`, () => {
    const isName = isObjectWithShape(
      {
        req: {
          firstName: isString,
          surname: isString,
        },
      },
      { name: "isName" }
    );
    const isPerson = isObjectWithShape(
      {
        req: {
          name: isName,
        },
      },
      { name: "isPerson" }
    );

    expect(() =>
      isPerson({ name: { firstName: "John" } })
    ).toThrowErrorMatchingSnapshot();
  });

  test(`passing object with nested wrong property in opt throws error`, () => {
    const isName = isObjectWithShape(
      {
        req: {
          firstName: isString,
          surname: isString,
        },
        opt: {
          title: isString,
        },
      },
      { name: "isName" }
    );
    const isPerson = isObjectWithShape(
      {
        req: {
          name: isName,
        },
      },
      { name: "isPerson" }
    );

    expect(() =>
      isPerson({ name: { firstName: "John", surname: "Smith", title: false } })
    ).toThrowErrorMatchingSnapshot();
  });
});
