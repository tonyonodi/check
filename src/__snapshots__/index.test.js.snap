// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isArray passing non-array throws a CheckError 1`] = `
"CheckError: check \\"isArray\\" expected an array but received 

	false

 at:
	isArray"
`;

exports[`isArrayOf passing array with element that fails test throws CheckError 1`] = `
"CheckError: check \\"isString\\" expected a string but received 

	2

 at:
	isString
	isArrayOfStrings[1]"
`;

exports[`isArrayOf passing non-array when expecting array throws CheckError 1`] = `
"CheckError: check \\"isArrayOfStrings\\" expected an array but received 5 at:
	isArrayOfStrings"
`;

exports[`isBool passing non-boolean falsey values throws a CheckError 1`] = `
"CheckError: check \\"isBool\\" expected a boolean but received 

	undefined

 at:
	isBool"
`;

exports[`isBool passing non-boolean falsey values throws a CheckError 2`] = `
"CheckError: check \\"isBool\\" expected a boolean but received 

	null

 at:
	isBool"
`;

exports[`isBool passing non-boolean falsey values throws a CheckError 3`] = `
"CheckError: check \\"isBool\\" expected a boolean but received 

	0

 at:
	isBool"
`;

exports[`isBool passing non-boolean truthy values throws a CheckError 1`] = `
"CheckError: check \\"isBool\\" expected a boolean but received 

	\\"a\\"

 at:
	isBool"
`;

exports[`isBool passing non-boolean truthy values throws a CheckError 2`] = `
"CheckError: check \\"isBool\\" expected a boolean but received 

	1

 at:
	isBool"
`;

exports[`isBool passing non-boolean truthy values throws a CheckError 3`] = `
"CheckError: check \\"isBool\\" expected a boolean but received 

	[]

 at:
	isBool"
`;

exports[`isInstanceOf passing incorrect value throws CheckError 1`] = `
"CheckError: check \\"isInstanceOf(Date)\\" expected an instance of instance of Date but got
      
  5
  
   at:
	isInstanceOf(Date)"
`;

exports[`isInteger passing float throws a CheckError 1`] = `
"CheckError: check \\"isInteger\\" expected an integer but received 

	5.1

 at:
	isInteger"
`;

exports[`isInteger passing string throws a CheckError 1`] = `
"CheckError: check \\"isInteger\\" expected an integer but received 

	\\"f\\"

 at:
	isInteger"
`;

exports[`isNumber passing a string throws a CheckError 1`] = `
"CheckError: check \\"isNumber\\" expected a number but received 

	\\"s\\"

 at:
	isNumber"
`;

exports[`isObjectWithShape passing object missing required key throws CheckError 1`] = `
"CheckError: check \\"isPerson.dateOfBirth\\" expected a property \\"dateOfBirth\\" at:
	isPerson.dateOfBirth"
`;

exports[`isObjectWithShape passing object with nested missing property in req throws error 1`] = `
"CheckError: check \\"isName.surname\\" expected a property \\"surname\\" at:
	isName.surname
	isPerson.name"
`;

exports[`isObjectWithShape passing object with nested wrong property in opt throws error 1`] = `
"CheckError: check \\"isString\\" expected a string but received 

	false

 at:
	isString
	isName.title
	isPerson.name"
`;

exports[`isObjectWithShape passing object with nested wrong property in req throws error 1`] = `
"CheckError: check \\"isString\\" expected a string but received 

	5

 at:
	isString
	isName.surname
	isPerson.name"
`;

exports[`isObjectWithShape passing object with wrong property in opt throws error 1`] = `
"CheckError: check \\"isInteger\\" expected an integer but received 

	199.5

 at:
	isInteger
	isPerson.dateOfBirth"
`;

exports[`isObjectWithShape passing object with wrong property in req throws error 1`] = `
"CheckError: check \\"isString\\" expected a string but received 

	5

 at:
	isString
	isPerson.firstName"
`;

exports[`isString passing an object throws a CheckError 1`] = `
"CheckError: check \\"isString\\" expected a string but received 

	{\\"a\\":5}

 at:
	isString"
`;

exports[`isVal passing wrong number throws CheckError 1`] = `
"CheckError: check \\"isVal(5)\\" expected value

  5
  
but received

  6
  
 at:
	isVal(5)"
`;

exports[`isVal passing wrong value when expecting array throws CheckError 1`] = `
"CheckError: check \\"isVal()\\" expected value

  []
  
but received

  5
  
 at:
	isVal()"
`;

exports[`isVal passing wrong value when expecting array throws CheckError 2`] = `
"CheckError: check \\"isVal(1,2,3)\\" expected value

  [1,2,3]
  
but received

  [1,2,2]
  
 at:
	isVal(1,2,3)"
`;

exports[`isVal passing wrong value when expecting nested array throws CheckError 1`] = `
"CheckError: check \\"isVal(1,2,3)\\" expected value

  [1,2,[3]]
  
but received

  [1,2,[4]]
  
 at:
	isVal(1,2,3)"
`;

exports[`isVal passing wrong value when expecting object throws CheckError 1`] = `
"CheckError: check \\"isVal([object Object])\\" expected value

  {}
  
but received

  5
  
 at:
	isVal([object Object])"
`;

exports[`isVal passing wrong value when expecting object throws CheckError 2`] = `
"CheckError: check \\"isVal([object Object])\\" expected value

  {}
  
but received

  {\\"a\\":1}
  
 at:
	isVal([object Object])"
`;

exports[`isVal passing wrong value when expecting object throws CheckError 3`] = `
"CheckError: check \\"isVal([object Object])\\" expected value

  {\\"num\\":1,\\"str\\":\\"string\\"}
  
but received

  {\\"a\\":1}
  
 at:
	isVal([object Object])"
`;

exports[`isVal passing wrong value when expecting object throws CheckError 4`] = `
"CheckError: check \\"isVal([object Object])\\" expected value

  {\\"num\\":1,\\"arr\\":[1,\\"b\\"],\\"obj\\":{\\"a\\":1}}
  
but received

  {\\"num\\":1,\\"arr\\":[1,\\"b\\"],\\"obj\\":{\\"a\\":2}}
  
 at:
	isVal([object Object])"
`;

exports[`maybe passing value with incorrect type throws CheckError 1`] = `
"CheckError: check \\"isString\\" expected a string but received 

	5

 at:
	isString"
`;
